File: Stack/Python/155_min_stack.py
# RUNTIME: 4 ms, faster than 71.82% of python3 Submissions. 
# MEMORY USAGE: 21.19 MB, less than 97.94% of python3 Submissions.


# My thougtprocess:
# 1. Define two lists - one stack - one min value
# 2. if stack is empty append min value
# 3. else append only to minValue if last min is less
# 4. pop stack from both lists if equal
# 5. return latest element in called lists


class MinStack:


    def __init__(self):
        self.stack = []
        self.minVal = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if (not self.minVal):
            self.minVal.append(val)
        elif (val <= self.minVal[-1]):
            self.minVal.append(val)

    def pop(self) -> None:
        if (self.stack[-1] == self.minVal[-1]):
            self.minVal = self.minVal[:-1]
        self.stack = self.stack[:-1]

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minVal[-1]

File: Stack/C++/155_min_stack.cpp
/*
* RUNTIME: 0 ms - Beats 100.00% of submissions  - don´t know how 0 ms is even possible
* MEMORY - 23.54 MB - Beats 32.74% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
#include <stack>
using namespace std;

/* Thoughtprocess:
* 1. initailize stack for input value and min value
* 2. if min-stack is empty first input is registered
* 3. push back min value and input value
* 
*/




class MinStack {
public:
    stack<int> inputStack;
    stack<int> minValStack;

    MinStack() {

    }
    
    void push(int val) {
        this->inputStack.push(val);
        if (this->minValStack.empty())
            this->minValStack.push(val);
        else if (val <= this->minValStack.top())
            this->minValStack.push(val);
    }
    
    void pop() {
        if (this->inputStack.top() == this->minValStack.top())
            this->minValStack.pop();
        this->inputStack.pop();
    }
    
    int top() {
        return this->inputStack.top();
    }
    
    int getMin() {
        return this->minValStack.top();
    }
};